[영화 극장 선택] 다이스 1, 2단계 미션 제출합니다.

moondev03 opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->

안녕하세요, 크롱!
레벨 1에 이어 이번 미션도 잘 부탁드립니다! 🙇‍♂️

### Activity UI Test
Fragment를 사용하게 되면서, 이제 Activity에서 확인할 테스트는 프래그먼트가 제대로 올라갔는지를 확인하는 정도뿐이라
Activity UI 테스트 자체가 과연 의미가 있는지 의문이 들었습니다.

이번에는 하단 내비게이션 탭을 클릭했을 때 어떤 프래그먼트가 보여지는지를 확인하는 테스트 코드를 작성했는데,
Activity 테스트에서 Fragment의 타입을 확인하는 작업이 적절한지에 대한 고민이 생겼습니다.


### XML - DataBinding

상태를 반영하는 작업은 구현했지만, 클릭 이벤트 처리 부분은 적절한 방법이 떠오르지 않아 수정하지 못했습니다.
보통 클릭 이벤트에서는 presenter의 메서드를 호출하게 되는데, 이때 presenter를 interface 타입으로 variable에 정의하고 있으니 그냥 presenter를 View(XML)에서 참조해도 상관없을 지에 대해 크롱의 의견이 궁금합니다.

⸻


## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

https://github.com/user-attachments/assets/5945ccfa-dcf8-4f13-bce9-e93ea73d3b3a

moondev03 committed:
[영화 티켓 예매] 마이그레이션 작업

moondev03 committed:
docs: 구현 기능 목록 작성

moondev03 committed:
feat: DataBinding 설정 작업

moondev03 committed:
feat: androidx-fragment-ktx 의존성 추가

moondev03 committed:
feat: BaseFragment 추가

- DataBinding inflate 중복 코드

moondev03 committed:
feat: fragment test 의존성 추가

moondev03 committed:
feat: Base 코드 수정

- showToast 역할 BaseFragment로 이전

moondev03 committed:
refactor: 영화 목록 화면 & 영화 예매 화면 Fragment로 변경

- 테스트 코드 수정

Commit:
refactor: 좌석 선택 화면 & 예매 결과 화면 Fragment로 변경

Commit:
feat : 영화관 및 영화 스케쥴 모델 구현

moondev03 committed:
feat: 영화관 더미 데이터 생성

moondev03 committed:
feat: 영화관 선택 바텀 시트 다이얼로그 구현

moondev03 committed:
refactor: BaseFragment Binding null 처리

moondev03 committed:
refactor: 영화관 선택 화면 노출 기능 구현

Commit:
refactor: 결과 화면 Views 코드 개선

- 매서드 분리

Commit:
refactor: 영화 예매 결과 화면에 영화관 표시 되도록 수정

Commit:
refactor: 더미 생성 로직 개선

Commit:
test: 테스트 코드 수정

moondev03 committed:
test: Presenter 테스트 코드 수정

moondev03 committed:
test: 도메인 모델 테스트 추가 및 수정

- 영화관, 영화관 목록, 영화 시간표 테스트 코드 추가
- 상영일 테스트 수정

moondev03 committed:
docs: 구현 기능 목록 갱신

moondev03 committed:
refactor: 영화 목록 화면 View 로직 분리

- Views 객체로 분리

moondev03 committed:
fix: 뒤로가기할 때 영화 목록 안그려지는 문제 해결

Commit:
feat: 하단 네비게이션 구현

moondev03 committed:
refactor: 예약 플로우를 위한 Activity 추가

moondev03 committed:
refactor: View 패키지 구조 변경

- view -> history, home, setting로 분리

moondev03 committed:
fix: Activity SaveInstanceState Null일 때만 Fragment 생성하도록 수정

moondev03 committed:
refactor: ReservationActivity 가독성 개선

moondev03 committed:
refactor: 파일 위치 수정

Commit:
refactor: TheaterBottomSheetDialogFragment MVP 적용

moondev03 committed:
refactor: 영화 예매 Activity 뒤로가기 로직 개선

moondev03 committed:
chore: origin/step1 -> step1에 병합

Commit:
test: TheaterBottomSheetDialogPresenter 테스트 코드 작성

moondev03 committed:
test: MoviesTheater UI 테스트 코드 작성

moondev03 committed:
refactor: 불필요한 코드 제거

moondev03 committed:
refactor: 영화 예매 결과 화면 DataBinding 적용

Commit:
refactor: 영화 목록 화면 DataBinding 적용

Commit:
refactor: 좌석 선택 화면 DataBinding 적용

Commit:
refactor: 좌석 선택 화면 개선

moondev03 committed:
refactor: 영화 예매 화면 DataBinding 적용

moondev03 committed:
fix: Movie 아이템 오류 수정

- IllegalFormatConversionException

Commit:
test: 테스트 수정 및 테스트 이름 변경

Commit:
test: 테스트 수정

moondev03 committed:
refactor: 영화 예매 좌석 선택 화면 개선

- Seat View 생성 로직 분리
- Presenter, View 수정

moondev03 committed:
fix: 영화 예매 좌석 선택 화면 화면 재구성 시 데이터 복구 수정

krrong commented on app/src/main/java/woowacourse/movie/domain/model/cinema/screen/Seat.kt:
```diff
@@ -0,0 +1,7 @@
+package woowacourse.movie.domain.model.cinema.screen
+
+data class Seat(
```
`Seat`은 왜 `Screen` 패키지 아래 있는 것인가요?

moondev03 replied:
현재 `Screen` 객체는 `영화관 좌석 구성 정보`만을 관리하고 있습니다.
이 때문에 `Seat`와 `Screen`을 같은 패키지에 두었더니 헷갈리게 한 것 같네요 😭
패키지 구조와 객체 이름을 더 적절하게 수정해보겠습니다!

[3a555fc](https://github.com/woowacourse/android-movie-theater/pull/115/commits/3a555fc9c84c12e2982e58763817227395f8c814)

[d57c96f](https://github.com/woowacourse/android-movie-theater/pull/115/commits/d57c96fcacf12659fa788843066bd4fa231d4136)

krrong commented on app/src/main/java/woowacourse/movie/domain/model/cinema/DiceCinemaPricePolicy.kt:
```diff
@@ -0,0 +1,12 @@
+
+import woowacourse.movie.domain.model.cinema.screen.SeatType
+
+class DiceCinemaPricePolicy : PricePolicy {
```
합리적인 가격의 Dice 시네마 😆

krrong commented on app/src/main/java/woowacourse/movie/domain/model/cinema/ticket/TicketBundle.kt:
```diff
@@ -0,0 +1,28 @@
+
+class TicketBundle private constructor(
+    tickets: List<Ticket>,
+) {
```
`Ticket`은 단일 티켓을 의미하고 `count`와 같은 변수 대신 `TicketBundle`로 여러 장의 `Ticket`을 관리하는 것처럼 보입니다.
이렇게 구현함으로써 얻은 이점이 무엇이었는지 공유해줄 수 있을까요?

moondev03 replied:
원래는 `Ticket`이 단일 좌석을 명확하게 표현하고 `TicketBundle`이 여러 장의 티켓을 의미 있게 묶는 구조가  
도메인 규칙 적용과 확장성 측면에서 더 유리하다고 생각하여 이렇게 구현했지만 
지금 구조는 각 티켓의 좌석 정보만 다른 것 같아서 `Ticket`의 `Seat`를 `List<Seat>`로 관리하게 해도 충분할 것 같습니다.

[dbf2cb8](https://github.com/woowacourse/android-movie-theater/pull/115/commits/dbf2cb817c65b5fe8cfb2122ee28757ba143f242)

krrong commented on app/src/main/java/woowacourse/movie/domain/model/cinema/PricePolicy.kt:
```diff
@@ -0,0 +1,7 @@
+
+import woowacourse.movie.domain.model.cinema.screen.SeatType
+
+interface PricePolicy {
```
인터페이스 활용 좋네요 👍

krrong commented on app/src/main/java/woowacourse/movie/domain/model/movie/Poster.kt:
```diff
@@ -0,0 +1,11 @@
+package woowacourse.movie.domain.model.movie
+
+sealed class Poster {
+    data class Resource(
+        val resId: Int,
+    ) : Poster()
+
+    data class Url(
+        val url: String,
+    ) : Poster()
+}
```
궁금한 부분입니다.
다이스는 이미지의 리소스도 도메인이라고 보시는걸까요? 그렇다면 이유는 무엇인지도 알려주세요.

moondev03 replied:
저 역시 `Resource ID`가 도메인 계층에 있는 것은 적절하지 않다고 생각합니다.
하지만 현재 영화 포스터 이미지는 프로젝트 내부에서 관리되고 있기 때문에 `Resource ID`가 함께 내려오지 않으면 별도의 매핑 작업이 필요합니다. 그런데 지난 미션에서 보유한 영화 정보만으로는 포스터 이미지를 정확히 매핑하기에 부족하다고 판단 했었습니다.
그래서 `Resource ID`가 함께 전달된다면 해당 ID를 기반으로 포스터 이미지를 쉽게 매핑하거나 바로 사용할 수 있을 것이라고 기대해 해당 방식으로 구현했었습니다.

하지만 이제 Movie에 고유한 ID가 생겼으니 Movie 객체에서는 포스터 정보를 제거하고 이 ID를 활용해 포스터 이미지가 포함된 UiModel을 구성하는 것이 더 나은 구조일까요?

krrong replied:
제가 이 코멘트를 통해 상기시켜드리고 싶었던 부분은 이미지의 리소스가 도메인인가 아닌가에 대한 고민이었습니다.

현재 미션에서는 다이스가 말해주신 것처럼 매핑하기 위한 정보가 부족하다라고 느껴질 수 있다는 점에 동의합니다.
그래서 영화에 대한 포스터 정보를 매핑하는 방법은 여러가지가 있을 수 있고 이에 대해서는 리뷰를 남기지 않고 있습니다.

언급해주신 방법도 좋은 방법인 것 같아요 👍

krrong commented on app/src/main/java/woowacourse/movie/domain/model/reservation/ReservationInfo.kt:
```diff
@@ -0,0 +1,29 @@
+    private val _seats: MutableList<Seat> = mutableListOf()
+    val seats: List<Seat> get() = _seats.toList()
+
+    fun updateSeats(seat: Seat) {
+        if (_seats.contains(seat)) {
+            _seats.remove(seat)
+            return
+        }
+
+        require(reservationCount.value > _seats.count()) { INVALID_SEATS_SIZE_ERROR_MESSAGE.format(reservationCount.value, _seats.count()) }
+        _seats.add(seat)
+    }
```
`update`라는 모든 것을 할 수 있을 것 같은 함수명 아래에서 자리를 추가하고 제거하는 일을 동시에 진행하고 있습니다. 
그렇다면 무엇을 해야 할까요?

moondev03 replied:
`addSeat`, `removeSeat`, `hasSeat` 메서드로 분리하여
외부에서 `hasSeat` 결과에 따라 좌석을 추가하거나 제거하도록 수정하였습니다.

또한, `hasSeat` 결과와 관계없이 중복 추가될 가능성을 고려하여 `_seats`를 `mutableSet`으로 변경하였습니다.

[c79b569](https://github.com/woowacourse/android-movie-theater/pull/115/commits/c79b569d7e2c9fd6a5081c4012d4a8919d22b692)

krrong commented on app/src/main/java/woowacourse/movie/domain/model/reservation/ReservationCount.kt:
```diff
@@ -0,0 +1,19 @@
+
+    operator fun plus(other: Int): ReservationCount = ReservationCount(value + other)
+
+    operator fun minus(other: Int): ReservationCount = ReservationCount(value - other)
```
잘 만들어두었지만, 효과적으로 사용하고 있지 못한 것 같습니다.

moondev03 replied:
단순히 `reservationCount += updateCount` 해당 연산에만 사용되고 있습니다.
이 때 새로운 객체가 생성되면서 init 블럭에 위치한 검증이 수행됩니다.

피드백 주신 말씀이 `operator 함수`가 굳이 필요하지 않았다는 뜻인지,
아니면 잘 설계되었지만 현재 코드에서 충분히 활용되고 있지 않다는 뜻인지 헷갈립니다 😭

krrong replied:
연산자를 오버로딩하여 잘 구현해두셨습니다.

하지만 언급해주신 것처럼 `reservationCount += updateCount`연산에서만 사용하고 있기 때문에 더하기, 빼기 연산을 각각 오버로딩한 것의 의미가 사라진다는 내용으로 남긴 코멘트였습니다 😅

moondev03 replied:
아하, 이해했습니다!

현재는 `plus` 연산자 함수만 사용하고 있어 `minus` 메서드는 제거했습니다.
또한, 이전에 알려주신 내용을 참고하여 plus 메서드 호출 시 runCatching을 사용해 새**로운 객체 생성에 실패하면 기존 객체를 반환**하도록 수정했습니다.

추가로, 영화 예매 화면에서 최소/최대 인원일 때 버튼이 비활성화되도록 하기 위해 isMin()과 isMax() 메서드를 추가했습니다.

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/reservation/detail/ReservationDetailPresenter.kt:
```diff
@@ -0,0 +1,119 @@
+    private fun getAvailableShowTimes(currentTime: LocalDateTime): List<LocalDateTime> =
+        theater.getAvailableShowTimesFor(movie.id, currentTime)
+
+    private fun ReservationCount.isValid(): Boolean = value > ReservationCount.RESERVATION_MIN_COUNT
```
도메인이 해야 하는 일이 아닐까요?

조금 더 나아가보면 `ReservationCount`을 불변 객체로 만들어두었기 때문에 티켓의 수량을 감소하는 경우 자기 자신을 반환하도록 할 수도 있어보입니다.

moondev03 replied:
`ReservationCount` 안에 최소 여부를 체크 메서드를 추가했습니다!

[e97c700](https://github.com/woowacourse/android-movie-theater/pull/115/commits/e97c7007f7126a95b971c6e7fb30afd85d07306b)

krrong commented on app/src/test/java/woowacourse/movie/domain/model/cinema/ticket/TicketMachineTest.kt:
```diff
@@ -0,0 +1,50 @@
+import java.time.LocalDateTime
+
+class TicketMachineTest {
+    private val policy = DiceCinemaPricePolicy()
```
인터페이스로 만들어두었지만 테스트에서는 클래스를 사용하고 있네요 😅

moondev03 replied:
```kotlin
private val mockPolicy =
    mockk<PricePolicy> {
        every { calculatePrice(SeatType.S_CLASS) } returns 15000
        every { calculatePrice(SeatType.B_CLASS) } returns 12000
        every { calculatePrice(SeatType.A_CLASS) } returns 10000
    }
```

`mockk`를 사용하여 `Mock 가격 정책`을 적용하여 테스트를 진행하도록 수정했습니다!

[1f5021e](https://github.com/woowacourse/android-movie-theater/pull/115/commits/1f5021e621dc6ce99e62566d2f17d3743b768a90)

krrong commented on app/src/main/java/woowacourse/movie/presentation/extension/PosterExtension.kt:
```diff
@@ -0,0 +1,13 @@
+import android.widget.ImageView
+import woowacourse.movie.presentation.model.PosterUiModel
+
+fun PosterUiModel.setImage(view: ImageView) {
```
사용하지 않는 것처럼 보입니다.

moondev03 replied:
BindingAdapter 적용 과정에서 필요없어진 것 같습니다!
파일 제거했습니다 :)

krrong commented on app/src/main/java/woowacourse/movie/presentation/base/BaseFragment.kt:
```diff
@@ -0,0 +1,36 @@
+abstract class BaseFragment<T : ViewDataBinding>(
+    @LayoutRes private val layoutResId: Int,
+) : Fragment() {
+    private var mBinding: T? = null
```
`mBinding`은 의미하는 바가 있나요?
현재 `binding`의 형태를 보면 익숙한 친구가 하나 떠오를 겁니다. 네이밍도 따라가볼 수 있겠어요.

moondev03 replied:
원래 `_binding`을 사용하려 했는데 `ktlint`에 걸려 해당 형태로 작성하였습니다.
하지만 공식 문서에서도 언더바(_)를 사용하니 `ktlint`에 예외를 추가하여 무시하도록 하겠습니다!

[58ba981](https://github.com/woowacourse/android-movie-theater/pull/115/commits/58ba9812cfb18966dcd1d5c99087e7a264916465)

krrong commented on app/src/main/java/woowacourse/movie/presentation/base/BaseFragment.kt:
```diff
@@ -0,0 +1,36 @@
+        Toast.makeText(requireContext(), message, Toast.LENGTH_SHORT).show()
+    }
+
+    override fun onDestroyView() {
+        super.onDestroyView()
+        mBinding = null
+    }
```
`binding`에 대한 null처리 👍

```kotlin
super.onDestroyView()
mBinding = null

mBinding = null
super.onDestroyView()
```
둘은 어떤 차이가 있는지 [공식문서](https://developer.android.com/topic/libraries/view-binding#fragments)를 보고 공부해보고 저에게 알려주세요.

moondev03 replied:
### _binding = null -> super.onDestroyView()
권장되지 않는 방식으로 Fragment의 View가 완전히 파괴되기 전에 ViewBinding 객체의 참조를 해제하면
View와 관련된 정리 작업이 제대로 이루어지지 않아 메모리 누수의 원인이 될 수 있다고 합니다.

###  super.onDestroyView() -> _binding = null
공식 문서에서 권장하는 방식으로 `super.onDestroyView()`를 먼저 호출하여 Fragment의 View와 관련된 내부 정리 작업을
완료한 후 `mBinding = null`를 통해 ViewBinding 객체가 View에 대한 참조를 유지하지 않도록 하여 메모리 누수를 방지할 수 있습니다.


핵심은 Fragment의 뷰가 완전히 파괴되기 전에 ViewBinding 객체의 참조를 해제하면 뷰와 관련된 정리 작업이 제대로 이루어지지 않아 메모리 누수의 원인이 될 수 있다는 점인 것 같습니다.

krrong commented on app/src/main/java/woowacourse/movie/presentation/base/BaseActivity.kt:
```diff
@@ -1,16 +1,30 @@
+import androidx.databinding.ViewDataBinding
+import woowacourse.movie.R
+
+abstract class BaseActivity<T : ViewDataBinding>(
```
`BaseActivity`가 있는 것을 보니 어느정도 안드로이드에 익숙해보이는 것 같습니다.
다이스는 `BaseActivity`에 어떤 것을 정의하나요?

moondev03 replied:
평소에는 다음과 같은 기능들을 `BaseActivity`에 정의해 사용해왔습니다.
- 공통 스낵바/토스트 표시
- 로딩 다이얼로그 표시/숨김 
- ViewBinding 초기화 작업

이외의 기능들은 일부 Activity에서만 사용되는 경우가 있어서
모든 Activity에서 공통적으로 필요한 기능만 `BaseActivity`에 포함하는 것이 좋다고 생각합니다.

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/MovieTheaterActivity.kt:
```diff
@@ -0,0 +1,46 @@
+        super.onCreate(savedInstanceState)
+        setBottomNavigationItemClickListener()
+
+        if (savedInstanceState == null) {
+            binding.bottomNavigation.selectedItemId = R.id.menu_home
+        }
```
`savedInstanceState` 에 대한 조건문을 사용하는 이유가 무엇일까요?

또한, `selectedItemId`를 홈으로 고정하면 어떤 문제가 발생할 수 있을지 고민해볼까요?

moondev03 replied:
`savedInstanceState`에 대한 조건문을 사용하는 이유는 초기화 작업을 한 번만 수행하기 위함입니다.

예를 들어, 화면 회전과 같은 이유로 액티비티가 재생성될 경우 이 조건문이 없으면 초기화 작업이 다시 실행됩니다.
이로 인해 설정 화면이나 예매 내역 화면에서 재구성이 발생하더라도 홈 화면으로 이동하는 등의 원치 않는 동작이 발생할 수 있습니다.
따라서 `savedInstanceState`가 `null`일 때만 초기화 작업을 수행하도록 조건을 설정했습니다.

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/MovieTheaterActivity.kt:
```diff
@@ -0,0 +1,46 @@
+            if (isSameNavItem(item)) return@setOnItemSelectedListener false
+
+            when (item.itemId) {
+                R.id.menu_home -> navigateToScreen(MoviesFragment())
+                R.id.menu_history -> navigateToScreen(ReservationHistoryFragment())
+                R.id.menu_setting -> navigateToScreen(SettingFragment())
```
매번 탭을 누를 때마다 새로운 Fragment를 만들어주어야할까요?

moondev03 replied:
`Fragment` Instance를 `Lazy`로 선언해놓고 재사용하도록 수정했습니다!

추가로 `add` 이후 `tag`를 사용하여 `show/hide` 처리되도록 개선했습니다.

[d418137](https://github.com/woowacourse/android-movie-theater/pull/115/commits/d4181374d0853be72c9d5824a494a7f727e55785)

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/MovieTheaterActivity.kt:
```diff
@@ -0,0 +1,46 @@
+
+    private fun isSameNavItem(item: MenuItem): Boolean = binding.bottomNavigation.selectedItemId == item.itemId
+
+    private fun navigateToScreen(fragment: Fragment) {
+        supportFragmentManager.commit {
+            setReorderingAllowed(true)
+            replace(R.id.fragment_container_view, fragment)
+        }
```
아래 질문들에 대해 공부해보시고 저에게도 간단하게 알려주세요.
1. `setReorderingAllowed(true)` 이 친구는 왜 있는 것인가요?
2. `replace`와 `add`는 무엇이 다른가요?
3. `commit` 과 `commitNow`는 어떻게 다른가요?
4. `supportFragmentManager.beginTransaction()`를 사용했을 때와 지금은 어떤 차이가 있나요?

moondev03 replied:
### setReorderingAllowed(true) 이 친구는 왜 있는 것인가요? ([공식 문서](https://developer.android.com/guide/fragments/transactions?hl=ko#reordering))
`setReorderingAllowed(true)`를 설정하는 이유는 프래그먼트의 생명주기 호출 순서를 최적화하고
전환 애니메이션이 누락되지 않도록 보장하며 백스택 처리 시에도 일관된 동작을 유지하기 위해서입니다.


### replace와 add는 무엇이 다른가요?
`replace()`는 remove -> add 수행
`add()`는 add 수행


### commit 과 commitNow는 어떻게 다른가요?
`commit`은 메인 스레드의 다음 프레임에서 수행되지만 `commitNow`는 즉시 실행됩니다.
때문에 `addToBackStack`과 `commitNow`를 같이 사용 시 문제가 발생합니다.


### supportFragmentManager.beginTransaction()를 사용했을 때와 지금은 어떤 차이가 있나요?
`commit {}` 방식도 내부에서는 `beginTransaction`을 사용하는 것으로 확인했습니다.
성능 상 차이는 크게 없는 것 같고 람다를 사용하여 가독성 측면에서 개선되었다는 점이 있는 것 같습니다.
```kotlin 
val bt = supportFragmentManager.beginTransaction()
bt.add()
bt.addToBackStack(null)
bt.commit()
```

krrong replied:
잘 공부하신 것 같습니다.

현재 시점에서 `setReorderingAllowed(true)`이 필요한 이유는 무엇인가요?
없으면 예상대로 동작하지 않는 것일까요?

moondev03 replied:
현재로서는 없어도 정상적으로 동작하는 것으로 보입니다.

`setReorderingAllowed(true)`에 대한 사용 사례를 찾아보니, 새로운 Fragment의 `onCreate()`가 이전 Fragment의 `onDestroy()`보다 먼저 호출될 수 있도록 보장해주는 기능으로 보입니다. 주로 `FragmentResultListener`, `Shared ViewModel`, `postponeEnterTransition`처럼 **타이밍에 민감한 로직을 안정적으로 처리**해야 할 때 사용된다고 합니다. 
또한 여러 트랜잭션을 연속적으로 실행하거나, 복잡한 `replace`/`add`/`remove` 조합을 사용하는 경우에도 사용되는 것으로 보입니다.

하지만 현재 상황에서는 이러한 사례에 해당하지 않으므로 해당 설정은 제거해도 괜찮을 것 같습니다.
모두 제거하도록 하겠습니다 :)

혹시 제가 잘못 이해한 부분이 있다면 피드백 부탁드려요!

[e97e731](https://github.com/woowacourse/android-movie-theater/pull/115/commits/e97e731899caf51ee80413ef99adca1af2e773c0)

---
### 추가
지금 다시 보니 리사이클러뷰의 아이템이 보여지는게 `setReorderingAllowed(false)`로 했을 때는 완전히 아이템이 로드되지 않은 상태에서 보여지게 되는데 `setReorderingAllowed(true)`일 때는 로드된 다음이 보여지는 것 같습니다.

이미지나 리사이클러뷰를 사용하는 화면으로 전환할 때는 `setReorderingAllowed(true)`를 사용하는 것이 좋은 것 같습니다..!

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/reservation/seat/ReservationSeatContract.kt:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    interface View {
+        fun showScreen(
+            reservationInfo: ReservationInfoUiModel,
+            screen: ScreenUiModel,
+            selectedSeats: List<SeatUiModel>,
+        )
+
+        fun updateSeatState(selectedSeat: SeatUiModel)
+
+        fun notifyTotalPrice(totalPrice: Int)
+
+        fun notifyCanPublish(canPublish: Boolean)
```
`showXxx`, `updateXxx`, `notifyXxx` 셋은 어떻게 다른 의미로 사용되고 있는 것인가요?

moondev03 replied:
| 메서드 접두사 | 의미         | 예시                          | 책임             |
|---------------|--------------|-------------------------------|------------------|
| `showXxx`     | 보여준다     | `showScreen(...)`             | UI에 표시        |
| `updateXxx`   | 상태 갱신    | `updateSeatState(...)` <br>`updateTotalPrice(...)`       | 내부 상태 변경   |
| `notifyXxx`   | 알린다       | `notifyCanPublish(...)` | 외부 통지 / 이벤트 트리거 |

위와 같은 의미로 사용하고 있습니다.
다만 리팩토링 과정에서 일관성을 가지지 않고 작성한 것 같아 수정했습니다!

[2f43455](https://github.com/woowacourse/android-movie-theater/pull/115/commits/2f434552e70e5b598148a332f55d0bfa82691a04)

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/reservation/seat/ReservationSeatPresenter.kt:
```diff
@@ -0,0 +1,75 @@
+class ReservationSeatPresenter(
+    private val view: ReservationSeatContract.View,
+) : ReservationSeatContract.Presenter {
+    private val machine = TicketMachine(DiceCinemaPricePolicy())
```
의도한 것인지 모르겠으나 인터페이스로 추상화한 이유가 전혀 없어진 것처럼 보이네요..

moondev03 replied:
```Kotlin
class ReservationSeatPresenter(
    private val view: ReservationSeatContract.View,
    policy: PricePolicy = DiceCinemaPricePolicy(),
) : ReservationSeatContract.Presenter {
    private val machine = TicketMachine(policy)
```

주입 받도록 수정했습니다!

krrong commented on app/src/main/java/woowacourse/movie/presentation/model/SeatUiModel.kt:
```diff
@@ -0,0 +1,19 @@
+    fun toLabel(): String = "${'A' + row}$col"
+}
+
+fun Seat.toUiModel(): SeatUiModel = SeatUiModel(row, col, type.toUiModel())
+
+fun SeatUiModel.toModel(): Seat = Seat(row, col, type.toModel())
```
nit: `toUiModel`, `toModel`은 한 끗 차이라 매우 졸린 상황이라면 헷갈릴 수도 있겠어요.

moondev03 replied:
toModel -> toDomain으로 수정했습니다!

[6362589](https://github.com/woowacourse/android-movie-theater/pull/115/commits/6362589ec5e0f54330ff1862fe578a9d1239c29d)

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/reservation/seat/ReservationSeatFragment.kt:
```diff
@@ -0,0 +1,122 @@
+    ReservationSeatContract.View {
+    private lateinit var presenter: ReservationSeatPresenter
+    private lateinit var views: ReservationSeatViews
+    private lateinit var publishDialogInfo: DialogInfo
+    private val dialog: CustomAlertDialog by lazy { CustomAlertDialog(requireContext()) }
```
이러한 정보를 프로퍼티로 가지고 있어야 하는 이유는 무엇일까요 ?.?

moondev03 replied:
메서드 안에서 생성하여 사용하게 되면 메서드가 호출될 때마다 새로운 객체가 생성되어야 하는데
미리 생성해두고 해당 객체를 사용하게 되면 객체 생성 비용을 아낄 수 있다는 장점이 있을 것 같았습니다.

AlretDialog 생성에서는 굳이 필요 없는 작업이었을까요?

krrong replied:
아하 그렇군요.
해당 부분은 정답은 없는 것 같아서 다이스의 의견을 가져가셔도 좋을 것 같습니다 🤠

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/movies/adapter/MoviesDiffUtil.kt:
```diff
@@ -0,0 +1,16 @@
+import androidx.recyclerview.widget.DiffUtil
+import woowacourse.movie.presentation.model.MovieUiModel
+
+object MoviesDiffUtil : DiffUtil.ItemCallback<MovieUiModel>() {
+    override fun areItemsTheSame(
+        oldItem: MovieUiModel,
+        newItem: MovieUiModel,
+    ): Boolean = oldItem.id == newItem.id
+
+    override fun areContentsTheSame(
+        oldItem: MovieUiModel,
+        newItem: MovieUiModel,
+    ): Boolean = oldItem == newItem
+}
```
`DiffUtil`에서 단순히 `==`으로만 비교하는 것은 충분하지 않습니다.
다른 예제들을 더 찾아보면 좋을 것 같습니다.

moondev03 replied:
대부분의 예제에서는 `areItemsTheSame`에서 고유한 값을 비교하고 
`areContentsTheSame`에서는 == 연산자를 사용하는 것으로 보였습니다..

UI와 관련된 상태 값이 포함되어 있지 않다면 == 만으로도 충분하다고 생각했는데
어떤 작업을 더 추가해야 할지 감이 잘 오지 않습니다 😔

krrong replied:
아 이 부분은 새벽에 확인을 하느라 id를 확인하지 못했던 것 같습니다 🥲
혼란을 드려 죄송하네요.

잘 구현해주신 것처럼 보입니다!

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/movies/adapter/MoviesAdapter.kt:
```diff
@@ -0,0 +1,60 @@
+
+    override fun getItemCount(): Int = currentList.size + getAdCount()
+
+    private fun isAdPosition(position: Int): Boolean = (position + 1) % AD_FREQUENCY == 0
+
+    private fun getMovieForPosition(position: Int): MovieUiModel? {
+        val realPosition = position - (position / AD_FREQUENCY)
+        return currentList.getOrNull(realPosition)
+    }
```
3번째마다 광고가 나오는 것은 도메인 로직일까요 아닐까요?

moondev03 replied:
`광고 삽입 주기`나 `노출 여부`가 비즈니스적으로 중요한 역할을 한다면 도메인 로직으로 볼 수 있겠지만
이번 미션에서의 광고 삽입 주기는 UI 또는 UX에 따른 표현 방식에 가깝기 때문에 
저는 뷰에서 충분히 처리할 수 있는 로직이라고 생각합니다.

다만 현재 처럼 포지션에 의존하는 것 보다는 외부에서 처리하고 넣어주는 것이 더 좋은 방향일 것 같아서
수정해봤습니다.

[dcf70e6](https://github.com/woowacourse/android-movie-theater/pull/115/commits/dcf70e6796245245228abbdf365a75ee04d688b6)

krrong replied:
만약 해당 도메인을 콘솔이나 웹으로 옮겨간다면 3번째마다 광고가 나오는 것에 대해 일일히 대응해야 할까요?
만약 요구사항이 변경되어 5번째마다 광고가 나오게 된다면 앱, 웹, 콘솔 모두를 변경해야 하는 것이 맞을까요?

moondev03 replied:
광고 삽입 정책을 구현해보면서 궁금한 점이 생겼습니다.
현재 구현 방식은 **영화 목록과 광고 데이터를 각각 받아서 영화 목록에 정책에 따라 광고 데이터를 삽입하는 구조**입니다.

그런데 생각해보니 **영화 데이터를 내려받을 때 광고 데이터를 포함해서 하나의 목록으로 내려받는 방식**도 있을 것 같아서, 
지금의 구현 방식이 적절한지 의문이 들었습니다.

이에 대한 크롱의 의견이 궁금합니다.

[ad05282](https://github.com/woowacourse/android-movie-theater/pull/115/commits/ad052822eab7df0d3cf2c1f9c0a29767b64f3d77)

krrong replied:
말씀해주신 부분도 충분히 고려해볼만한 점이에요. 외부에서 데이터를 어떻게 받아올 것이냐도 구조를 결정하는데 큰 역할을 할겁니다.
하지만 현재 시점에서는 현재 가지고 계신 의문점만 가지고 가도 충분해보입니다. 🤠

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/movies/MoviesFragment.kt:
```diff
@@ -0,0 +1,54 @@
+    ) {
+        super.onViewCreated(view, savedInstanceState)
+
+        val moviesAdapter =
+            MoviesAdapter(
+                object : OnMovieEventListener {
+                    override fun onClick(movie: MovieUiModel) {
+                        presenter.availableTheatersAndCount(movie.id)
+                    }
+                },
+            )
```
무명 객체를 만들지 않도록 수정해볼 수 있을까요? 그럼 어떤 이점이 있을까요?

moondev03 replied:
말씀해주신 내용이 `무명 객체` 대신 `별도의 클래스로 분리`의 의미인지
아니면 `Lambda `나 `fun interface`를 활용하라는 의미인지 다소 헷갈리긴 했지만,  
일단 저는 이벤트가 늘어남에 따라 생기는 코드의 가독성 저하에 때문에 `무명 객체` 대신 `별도의 클래스로 분리`하여 구현해 보았습니다.

[760f27b](https://github.com/woowacourse/android-movie-theater/pull/115/commits/760f27b930f6bb1637d55f418a57153d59ae225f)

추가로 `Lambda`나 `fun interface`를 사용하면 더 간단하게 처리할 수 있지만
향후 `Long Click`, `Double Click`, `Drag` 등 다양한 이벤트가 추가될 가능성을 고려한다면,  
처음부터 `무명 객체`나 `별도의 클래스`로 구현해 두는 것이  `확장성`과 `유지보수` 측면에서 더 유리하다고 판단했습니다.

다만, 이번 미션처럼 이벤트가 단순하고  추가될 가능성이 거의 없는 경우에는  
오히려 `무명 객체`나 `별도의 클래스`보다는  `Lambda`나 `fun interface`를 사용하는 것이  
더 `적절한 선택`일 수도 있다는 생각도 듭니다.

이런 상황에서 `크롱`은 어떤 방향이 더 바람직하다고 생각하시나요?

krrong replied:
별도의 인터페이스로의 분리를 의도했었고 잘 구현해주신 것 같습니다.
인터페이스로 분리하게되면 다이스가 언급해주신 것처럼 다양한 이벤트가 생겨도 어렵지 않게 대응이 가능할거에요.

그런데 무명 객체로 만드는 것이 확장성과 유지보수 측면에서 더 유리하다고 생각한 이유가 무엇인지 설명해주실 수 있나요?

moondev03 replied:
제가 익명 객체가 더 확장성과 유지보수에 유리하다고 말씀드린 것은 람다나 함수형 인터페이스와 비교했을 때 그렇다는 의미였습니다.

람다 방식은 코드가 간결해지는 장점이 있지만 이벤트 마다 람다를 추가해야 하고,
함수형 인터페이스는 하나의 이벤트만 처리할 수 있다는 제한이 있습니다.
이런 단점을 익명 객체나 별도의 클래스로 분리하는 것으로 보완할 수 있다고 생각했습니다.

저 역시 가장 좋은 방식은 역할에 따라 코드를 별도로 분리하는 것이라고 생각합니다!

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/movies/adapter/AdViewHolder.kt:
```diff
@@ -0,0 +1,19 @@
+import androidx.recyclerview.widget.RecyclerView
+import woowacourse.movie.R
+
+class AdViewHolder(
+    parentView: View,
+) : RecyclerView.ViewHolder(parentView) {
+    private val ivAd = parentView.findViewById<ImageView>(R.id.iv_ad)
```
`MovieViewHolder`와 다르게 `findViewById`를 사용하고 있는데 특별한 이유가 있을까요?

moondev03 replied:
실수로 넘어간 것 같습니다 ㅠㅠ
Databinding 사용하도록 수정했습니다!

[c2e29cb](https://github.com/woowacourse/android-movie-theater/pull/115/commits/c2e29cbe42e8473e586f1b93fcd16c43b9624837)

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/movies/adapter/MovieBindingAdapter.kt:
```diff
@@ -0,0 +1,18 @@
+package woowacourse.movie.presentation.view.home.movies.adapter
```
Binding Adapter도 Adapter 이긴 하지만 성격이 조금 다르다고 생각합니다. 패키지를 분리해보는 것은 어떻게 생각하시나요? 정답은 없기 때문에 다이스의 의견을 남겨주셔도 됩니다.

moondev03 replied:
저도 별도로 분리하는게 맞다는 생각이 듭니다.
binding이라는 패키지를 생성하여 이동시켰습니다.

krrong commented on app/src/main/java/woowacourse/movie/presentation/view/home/movies/dialog/TheatersAdapter.kt:
```diff
@@ -0,0 +1,30 @@
+        return TheaterViewHolder(binding, onClickTheater)
+    }
+
+    override fun onBindViewHolder(
+        holder: RecyclerView.ViewHolder,
+        position: Int,
+    ) {
+        (holder as? TheaterViewHolder)?.bind(currentList[position])
+    }
```
`ViewHolder` 제네릭 타입에 무엇을 넣을 수 있는지 더 찾아보면 좋을 것 같습니다.

moondev03 replied:
여러 ViewType을 사용하지 않는 경우
`RecyclerView.ViewHolder` 대신 `TheaterViewHolder`로 넣어주어 타입 캐스팅이 필요없어졌습니다!

[9f7a224](https://github.com/woowacourse/android-movie-theater/pull/115/commits/9f7a22458819dd18cbb90349af8f9e0647342f7c)

krrong reviewed (changes requested):
안녕하세요 다이스, 다시 만나게 되었네요 반갑습니다 😆
이번에는 더욱 발전된 모습을 볼 수 있을 것 같은데 기대해도 되겠죠?

구현해주신 것을 보니 안드로이드에 익숙한 것처럼 보입니다.
도메인에 대한 부분부터 확인하고 넘어가면 좋을 것 같아 도메인부터 시작했습니다.

질문 주신 부분은 앞으로 리뷰를 주고 받으면서 더 이야기 해보도록 해요.
궁금하거나 어려운 부분은 언제든 연락주시면 됩니다 🙇‍♂️

moondev03 committed:
refactor: MovieTheaterActivity 바텀 내비게이션 전환 로직 개선

- add 이후 show/hide 처리
- Fragment Instance 생성 후 재사용

moondev03 committed:
feat: BaseViewHolder & BaseViewHolderItem 구현

moondev03 committed:
refactor: MoviesAdapter 개선

- 아이템 ViewType 분리
- 중복 코드 줄임
- 광고 삽입 로직 제거

moondev03 committed:
refactor: 영화 목록 광고 삽입 로직 분리

moondev03 committed:
refactor: BindingAdapter 개선

- 패키지 위치 이동
- 리스트에 아이템 목록 제출 전 광고 삽입 하도록 수정

moondev03 committed:
refactor: MoviesAdapter EventListener 별도의 클래스로 분리

- 익명 객체 대신 별도의 클래스로 분리

moondev03 committed:
refactor: 패키지 구조 변경

moondev03 committed:
refactor: Theater ListAdapter RecyclerView.ViewHolder 대신 TheaterViewHolder로 대체

moondev03 committed:
style: BaseFragment ViewBinding 변수 네이밍 수정

- mBinding -> _binding

moondev03 committed:
refactor: 광고 이미지 BindingAdapter 사용하도록 수정

- 기타 개선

moondev03 committed:
refactor: TicketBundle 제거

moondev03 committed:
refactor: DateTimeFormat BindingAdapter 구현

moondev03 committed:
refactor: Screen - Seat 구조 개선

- 역할 재분배

moondev03 committed:
refactor: Domain Model 패키지 구조 변경 및 네이밍 변경

moondev03 committed:
refactor: 영화관 선택 다이얼로그 DataBinding 적용

moondev03 committed:
refactor: 영화 시간표 더미 생성 로직 분리

moondev03 committed:
refactor: toModel -> toDomain으로 변경

moondev03 committed:
fix: 테스트 코드 버그 수정

moondev03 committed:
refactor: 티켓 가격 계산 테스트 개선

- 불필요한 테스트 코드 제거

moondev03 committed:
refactor: 예매 인원 개선

- 최소 예매 인원 여부 확인 메서드 추가

moondev03 committed:
refactor: 선택된 좌석 추가/제거 로직 분리

- 추가/제거 로직으로 분리
- hasSeat 메서드 추가

moondev03 committed:
refactor: View 네이밍 변경

- showXxx / updateXxx / notifyXxx

moondev03 committed:
refactor: 하단 내비게이션 메뉴 문자열 정리

moondev03 committed:
refactor: ReservationSeatPresenter 가독성 개선

moondev03 commented:
@coderabbitai pause

moondev03 committed:
refactor: 오탈자 수정

moondev03 reviewed (commented):
코드 래빗 리뷰도 확인했고 그중 몇 가지는 반영했지만 도움이 되는 리뷰도 많았지만 그렇지 않은 리뷰도 많았던 것 같습니다. 

그리고 코드 래빗이 남긴 코멘트 수가 너무 많아서 그런지 코멘트를 달 때마다 렉이 걸리는 느낌이 들었고
정작 크롱의 리뷰를 확인하기가 어려워 코드 래빗의 리뷰는 일단 비활성화 해두었습니다.

다만 코드 래빗의 리뷰가 필요하다고 판단되신다면 말씀해주세요. 
다시 활성화하겠습니다!

krrong commented on app/src/main/java/woowacourse/movie/presentation/home/movies/adapter/MoviesAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+
+    override fun getItemCount(): Int = currentList.size
+
+    interface OnMovieEventListener : MovieViewHolder.OnMovieEventListener
```
이 구현의 의도는 무엇일까요..?

moondev03 replied:
해당 인터페이스는 MoviesAdapter에서 사용된 ViewHolder들의 리스너를 외부로 모두 전달할 수 있도록 구현한 것이었습니다.
하지만 현재는 광고 클릭에 대한 요구사항이 없어, 불필요한 구현인 것 같아 제거하도록 하겠습니다..!

[19ad5f8](https://github.com/woowacourse/android-movie-theater/pull/115/commits/19ad5f8633c5e61eabf501aeac6600e766b344f7)

krrong commented on app/src/main/java/woowacourse/movie/presentation/home/movies/adapter/MoviesAdapter.kt:
```diff
@@ -0,0 +1,36 @@
+        when (MovieMainItem.MovieViewType.entries[viewType]) {
+            MovieMainItem.MovieViewType.MOVIE -> MovieViewHolder(parent, eventListener)
+            MovieMainItem.MovieViewType.AD ->
+                woowacourse.movie.presentation.home.movies.adapter.AdViewHolder(
```
nit: import를 했기 때문에 지워도 될 부분이 보입니다.
전체적으로 확인해주세요.

krrong commented on app/src/main/java/woowacourse/movie/presentation/common/base/BaseViewHolder.kt:
```diff
@@ -0,0 +1,10 @@
+import androidx.recyclerview.widget.RecyclerView
+import androidx.viewbinding.ViewBinding
+
+abstract class BaseViewHolder<T, B : ViewBinding>(
+    protected val binding: B,
+) : RecyclerView.ViewHolder(binding.root) {
+    abstract fun bind(item: T)
+}
```
현재 미션에서 정말 `BaseViewHolder`가 필요하다고 생각하시나요?
이 때문에 불필요한 IDE 경고를 만나고 있지는 않나요?

만약 답변이 확장성때문에 그렇습니다라고 한다면 그것은 더 고민해봐야 하는 답변입니다.
이는 우리가 구현하는 모든 것에 확장성을 고려해야 한다. 즉, 모든 클래스는 상위 클래스든 인터페이스는 무엇이든 구현하고 있어야 한다는 말과 같습니다.

moondev03 replied:
캠퍼스 내에서 진행한 미니 세션에서 다룬 내용을 바탕으로 이해되는 내용까지 구현해본 코드였습니다.

BaseViewHolder를 구현함으로써 ViewHolder가 추가 혹은 제거될 때 Adapter에서의 코드 수정을 줄일 수 있다는 점에서
유지 보수 측면에서 좋다는 생각에 적용해도 좋을 것 같다는 생각을 했었습니다.

앞으로는 정말 필요한 지에 대해 더 고민해보고 적용해보도록 하겠습니다 😔

krrong commented on app/src/main/java/woowacourse/movie/presentation/home/movies/adapter/MovieViewHolder.kt:
```diff
@@ -0,0 +1,26 @@
+    view: ViewGroup,
+    private val eventListener: OnMovieEventListener,
+) : BaseViewHolder<MovieItem, ItemMovieBinding>(
+        DataBindingUtil.inflate(LayoutInflater.from(view.context), R.layout.item_movie, view, false),
```
`Adapter`와 `ViewHolder`의 역할을 잘 나누어주셨네요 👍

krrong commented on app/src/main/java/woowacourse/movie/presentation/home/movies/adapter/MovieViewHolder.kt:
```diff
@@ -0,0 +1,26 @@
+    override fun bind(item: MovieItem) {
+        binding.itemMovie = item.movie
+        binding.onClickListener = eventListener
+    }
```
`bind` 함수가 호출될 때마다 리스너를 연결시켜줄 필요가 있을까요?

moondev03 replied:
init 블록에서 연결해주도록 수정했습니다!

[a640b53](https://github.com/woowacourse/android-movie-theater/pull/115/commits/a640b53d76de7e4eac769f2f4b2ab0d6b7d5bb00)

-> 개선
[a1b748f](https://github.com/woowacourse/android-movie-theater/pull/115/commits/a1b748f8c91c24b959f3b1e330ff84bd85bb86bb)

krrong commented on app/src/main/java/woowacourse/movie/presentation/home/movies/MovieEventListener.kt:
```diff
@@ -0,0 +1,12 @@
+    override fun onClick(movie: MovieUiModel) {
+        presenter.availableTheatersAndCount(movie.id)
+    }
+}
```
잘 구현해주셨습니다만, 특정 presenter와 강하게 결합되어 있는 것처럼 보입니다.
어떻게 개선해볼 수 있을까요?

moondev03 replied:
해당 코드를 제거하고 Fragment에 Listener를 상속시켜 override 하도록 수정했습니다.
특정 presenter와 강하게 결합되는 문제를 없애기는 했지만 이 방법을 의도하신 건지에 대한 확신이 들지는 않는 것 같습니다 😔

[a1b748f](https://github.com/woowacourse/android-movie-theater/pull/115/commits/a1b748f8c91c24b959f3b1e330ff84bd85bb86bb)

krrong replied:
구현하신 방법도 좋은 해결방법처럼 보입니다.
만약 인터페이스에 정의되어 있는 함수가 많아진다면 어떻게 될지도 고민해보면 도움이 될 것 같아요.

제가 의도했던 바는 함수를 넘겨주는 방식이었어요 🤠

krrong commented on app/src/main/java/woowacourse/movie/presentation/common/base/BaseViewHolderItem.kt:
```diff
@@ -0,0 +1,8 @@
+    val viewType: Int
+
+    override operator fun equals(other: Any?): Boolean
+}
```
해당 인터페이스가 필요한 이유가 무엇인가요?

moondev03 replied:
구현 중에 == 연산이 안되는 문제 때문에 추가했다가 리팩토링 과정에서 필요없어진 것 같습니다.
제거 했습니다!

[1b97f9d](https://github.com/woowacourse/android-movie-theater/pull/115/commits/1b97f9dc773a6ae6be4ebb212bd95b0343950182)

krrong commented on app/src/main/java/woowacourse/movie/presentation/common/binding/ImageViewBindingAdapter.kt:
```diff
@@ -0,0 +1,10 @@
+@BindingAdapter("srcBitmap")
+fun ImageView.setImageViewResource(bitmap: Bitmap) {
+    this.setImageBitmap(bitmap)
+}
```
[공식문서](https://developer.android.com/topic/libraries/data-binding/binding-adapters)를 참고해보시면 보통 `BindingAdapter`를 작성할 때는 첫 번째 인자로 뷰를 넣어줍니다.

moondev03 replied:
찾아 보니 dataBinding 예전 버전에서는 확장 함수 형태가 지원되지 않았지만 
현재는 컴파일 시점에서 처리되도록 개선되어 가능하다고 합니다. 

하지만 공식 문서에서는 인자로 받는 형태이기 때문에 수정해보겠습니다!

[76d39e4](https://github.com/woowacourse/android-movie-theater/pull/115/commits/76d39e4034cdb23880afeac5f02044960f934dcd)

krrong commented on app/src/main/java/woowacourse/movie/presentation/common/binding/PosterBindingAdapter.kt:
```diff
@@ -0,0 +1,12 @@
+import woowacourse.movie.presentation.common.model.PosterUiModel
+
+@BindingAdapter("poster")
+fun ImageView.setImageViewResource(poster: PosterUiModel) {
```
같은 이름을 가지고 있고 같은 일을 하는 `BindingAdapter`와 함께 범용적으로 사용할 수 있도록 만들어볼 수 있을 것 같습니다.

moondev03 replied:
poster를 제거하고 기존 BindingAdapter를 활용하도록 수정했습니다!

[68523f5](https://github.com/woowacourse/android-movie-theater/pull/115/commits/68523f5a05d5ed3a48c1553e9d5f4b58dee55fb4)

krrong commented on app/src/main/java/woowacourse/movie/presentation/common/custom/CustomAlertDialog.kt:
```diff
@@ -0,0 +1,45 @@
+                dialogInfo.onClickNegativeButton(dialog)
+            }
+        }
+    }
```
`update` 함수를 조금 수정한다면 `return builder.update(dialogInfo).create()` 처럼 만들어 레벨 1에서 배운 DSL을 사용해볼 수도 있을 것 같습니다.

moondev03 replied:
Builder 의 확장함수로 수정해봤습니다.
추가로 `dialog가 null`이거나 `DialogInfo가 이전과 다르다면` **새로운 dialog를 생성**하도록 개선해봤습니다.

[42a8e72](https://github.com/woowacourse/android-movie-theater/pull/115/commits/42a8e7281d62e18bbba89804b9f810eeb8a5333b)

krrong commented on app/src/main/java/woowacourse/movie/presentation/common/custom/CustomAlertDialog.kt:
```diff
@@ -0,0 +1,45 @@
+            dialog.setButton(AlertDialog.BUTTON_NEGATIVE, it) { _, _ ->
+                dialogInfo.onClickNegativeButton(dialog)
+            }
+        }
```
빌더의 함수를 사용하지 않고 직접 구현한 이유가 있는 것일까요?.?

moondev03 replied:
미처 생각하지 못했던 것 같습니다.. 😔

[42a8e72](https://github.com/woowacourse/android-movie-theater/pull/115/commits/42a8e7281d62e18bbba89804b9f810eeb8a5333b)

krrong commented on app/src/main/java/woowacourse/movie/presentation/home/reservation/detail/ReservationDetailFragment.kt:
```diff
@@ -0,0 +1,194 @@
+import java.time.LocalDateTime
+import java.time.LocalTime
+
+class ReservationDetailFragment :
```
이제 확인했는데 대부분의 화면이 프래그먼트로 구성되어 있네요..?
이렇게 구성한 이유가 있을까요?

현재 시점에서는 해당 부분들을 프래그먼트로 구성해서 어떤 이점을 누리고 있는지 확인하기 어려워 보입니다.. 🤔

moondev03 replied:
필요성에 대한 고민 없이 단순히 Fragment 활용이라는 요구사항에 맞춰 메인과 예매 플로우로 나누어 생각하고
각각을 Activity로 만들어 기존의 예매, 좌석 선택, 결과 화면을 Fragment로 수정하여 Activity에 추가해야겠다고 판단했던 것 같습니다. 

앞으로는 필요성에 대해 더 고민한 후 구현해보겠습니다 :)

krrong commented on app/src/main/res/layout/fragment_reservation_detail.xml:
```diff
@@ -0,0 +1,175 @@
+                tools:src="@drawable/harrypotter" />
+
+            <LinearLayout
+                android:layout_width="match_parent"
+                android:layout_height="0dp"
+                android:layout_weight="1"
+                android:gravity="center"
+                android:orientation="vertical"
+                android:paddingHorizontal="20dp">
+
+                <TextView
+                    android:id="@+id/tv_reservation_title"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:text="@{movie.title}"
+                    android:textColor="@color/black"
+                    android:textSize="24sp"
+                    android:textStyle="bold"
+                    tools:text="해리 포터와 마법사의 돌" />
+
+                <TextView
+                    android:id="@+id/tv_screening_period"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_marginTop="29dp"
+                    android:text="@{String.format(@string/movie_date, movie.screeningPeriod.formatedStartDate(@string/movie_screening_period_format), movie.screeningPeriod.formatedEndDate(@string/movie_screening_period_format))}"
+                    android:textColor="@color/black"
+                    android:textSize="16sp"
+                    tools:text="상영일: 2025.4.1 ~ 2025.4.25" />
+
+                <TextView
+                    android:id="@+id/tv_reservation_running_time"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:text="@{String.format(@string/running_time, movie.runningTime)}"
+                    android:textColor="@color/black"
+                    android:textSize="16sp"
+                    tools:text="러닝타임: 152분" />
+
+                <LinearLayout
```
여러 겹으로 중첩된 `LinearLayout`을 사용하면 어떤 문제가 발생할 수 있을까요?

moondev03 replied:
중첩된 레이아웃이 많아질 수록 성능 이슈와 가독성이 저하되어 유지보수나 수정이 어려워 질 수 있을 것 같습니다.
ConstraintLayout으로 수정해보겠습니다!

[eb4436d](https://github.com/woowacourse/android-movie-theater/pull/115/commits/eb4436da434216c18aeaebd85ddf2cae4806467c)

krrong reviewed (changes requested):
안녕하세요 다이스 리뷰 반영 고생 많으셨습니다 👍
전반적으로 잘 반영해주신 것 같아요.

안드로이드에 익숙하신 것처럼 보이고 이로 인해 불필요한 것들이 많이 첨가되어 있어, 리뷰를 어렵게 한다는 생각이 들었습니다.
우테코에서 지향하는 것처럼 마른 스펀지가 되어보는 것은 어떨까 조심스레 의견을 드리고 싶네요.

고민해볼만한 부분에 리뷰를 남겨드렸으니 확인해주시고,
궁금하거나 어려운 부분은 언제든 연락주시면 됩니다 🙇‍♂️

moondev03 committed:
refactor: 광고 삽입 빈도 도메인 모델 구현

- 빈도를 presenter에서 주입 받아 사용
- AdInjector 구현

moondev03 committed:
refactor: ReservationCount 개선

- 불필요한 minus 메서드 제거
- isMax 메서드 추가

moondev03 committed:
refactor: 영화 예매 화면 코드 개선

- ReservationCountUiModel(count, isMin, isMax) 추가
- 불필요한 runCatch 제거
- DataBinding 개선
- 테스트 코드 수정

moondev03 committed:
refactor: setReorderAllowed 제거

moondev03 committed:
refactor: 광고 삽입 정책 구현

- Presenter에서 주입받아 사용

moondev03 committed:
refactor: 불필요한 이벤트 리스너 제거

- MoviesAdapter

moondev03 committed:
style: 불필요한 Path 사용 제거

moondev03 committed:
refactor: ViewHolder Listener 추가 시점 변경

moondev03 committed:
refactor: ViewHolder Listener 개선

moondev03 committed:
refactor: BaseViewHolderItem equals override 제거

moondev03 committed:
refactor: BindingAdapter 함수 형식 변경

- 뷰를 확장에서 인자로 받도록 수정

moondev03 committed:
refactor: ImageView BindingAdapter 통합

- Poster + Bitmap

moondev03 committed:
fix: TestFixture 수정

moondev03 committed:
refactor: CustomAlertDialog 코드 개선

- 기존 다이얼로그 정보가 다르면 갱신

moondev03 committed:
refactor: DialogInfo Data Class로 변경

- 비교를 위해

moondev03 committed:
refactor: 영화 예매 화면 ConstraintLayout으로 변경

moondev03 reviewed (commented):
리뷰가 pending 상태인 걸 지금 확인했네요 죄송합니다 😭

krrong reviewed (approved):
안녕하세요 다이스! 리뷰 반영 잘 해주셨습니다 👍

새로운 방법을 적용하는데 있어 정말 필요한 방법일지 한번 더 고민해보면 좋을 것 같습니다 🤠
또한, 리팩터링을 하시면서 조그마한 부분을 종종 놓치고 계신데 해당 부분을 조금 더 신경써보도록 하죠.

1, 2단계는 여기서 마무리하고 다음 단계로 넘어가보도록 해요.
다음 단계도 응원하겠습니다 🙇‍♂️
